### ISSUE TITLE
Trigonometric functions different results for numpy and jax.numpy

### ISSUE THREAD

**Issue by sjherzog** [google/jax#7019]
Dear Devs,

while porting a numerical simulation from numpy to jax.numpy i noticed that the trigonometric function of jax.numpy and numpy give different results. The error becomes smaller when using float64, but is still significant. Here is some source code where you can observe the phenomenon:

```python
from jax.config import config
config.update("jax_enable_x64", True)
import jax
import jax.numpy as jnp
import numpy as np


x_jax = jax.random.uniform(jax.random.PRNGKey(
    1337), (512, 512), dtype=jnp.float64)
x_np = np.asarray(x_jax)

print("dtype numpy:", x_np.dtype)
print("dtype jax:", x_jax.dtype)

assert (x_jax == x_np).all()

if(not (jnp.sin(x_jax) == np.sin(x_np)).all()):
    print("Error for sin:", jnp.sum(jnp.sin(x_jax) - np.sin(x_np)))

if(not (jnp.cos(x_jax) == np.cos(x_np)).all()):
    print("Error for cos:", jnp.sum(jnp.cos(x_jax) - np.cos(x_np)))

if(not (jnp.tanh(x_jax) == np.tanh(x_np)).all()):
    print("Error for tanh:", jnp.sum(jnp.tanh(x_jax) - np.tanh(x_np)))
```

I've been able to reproduce this on two different computers with 

jaxlib==0.1.67+cuda111 and jaxlib==0.1.65+cuda111. I'm not sure if this is a bug or if the behavior is intentional due to specific gpu implementations. In any case, I would not have expected that =)

Best 
Sebastian

**Comment by jakevdp** [comment]
In general I would not expect floating point computations of transendental functions to produce exactly equal results when computed on different platforms (here JAX on GPU vs. numpy on CPU). Rather than checking for exact equality, you might check for equality within floating point precision using something like `numpy.allclose`.

**Comment by hawkinsp** [comment]
On GPU, for `sin` and `cos`, XLA ultimately calls `nv_sinf` and `nv_cosf` which are accurate to 2 ULP (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#standard-functions).

I suggest using `np.testing.assert_array_almost_equal_nulp(jnp.sin(x_jax), np.sin(x_np), nulp=2)` to compare.

I note our `tanh` is slightly less accurate; if that turns out to be a problem we can look into it.

**Comment by sjherzog** [comment]
For me, the `tanh` would indeed be the relevant one

**Comment by hawkinsp** [comment]
@sjherzog can you say how many ulps of accuracy you need and why? Again expecting exact equality is not reasonable.

**Comment by sjherzog** [comment]
@hawkinsp Sure, I’m simulating a chaotic system to describe the electrophysiological behaviour of cardiac muscle cells. Since the system is chaotic, even the smallest errors add up to an exponentially increasing error. I would then like to use the simulation to find new patterns that allow for gentler defibrillation.

I understand that I should not expect to have the same accuracy between two architectures. But shouldn't  tanh with double precision have an ULP of 1? Or am I understanding this wrong? (Sorry if yes)

If the behaviour is intentional i can close the issue.

**Comment by hawkinsp** [comment]
@sjherzog No, even Nvidia's `tanh` is only accurate to 2 ULP. (See https://en.wikipedia.org/wiki/Unit_in_the_last_place for what ULP means, note that an ULP is a type and value dependent accuracy bound.)

For `float64` XLA calls NVidia's implementation. For `float32` XLA uses a faster and less accurate tanh on purpose, because `tanh` is a key operation for neural networks. See: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/service/gpu/elemental_ir_emitter.cc;l=283

By the way, do you have evidence NumPy's `tanh` is actually more accurate? You'd probably need to compare against a multiprecision implementation to establish that.

In general there is a tradeoff between accuracy and performance. It's likely the NVidia-provided kernel is probably a good tradeoff between both.

**Comment by sjherzog** [comment]
@hawkinsp ahh i misunderstood, thanks for that!

As for the comparison to a multiprecision implementation, i once compared the duration of the transient chaos that exists in the system with the duration from a fortran implementation of mine. The duration of the numpy simulation and the fortran simulation was the same, but in the jax.numpy version the errors accumulated too much and the duration of the transient chaos differs by orders of magnitude.

But i'll close the issue and thank you for the explanation!
